src/plonk/permutation.rs

```diff
@@ -129,6 +129,79 @@ impl PermutationData {
     }
+
+    #[instrument(level = "debug", skip_all)]
+    /// Removes copy constraints for specific instance columns from the permutation mapping.
+    ///
+    /// # Parameters
+    ///
+    /// - `instance_columns_to_remove`: An iterator over the indices of instance columns
+    ///   for which copy constraints need to be removed.
+    ///
+    /// # Returns
+    ///
+    /// A new `PermutationData` object with the specified copy constraints removed.
+    ///
+    /// # Algorithm
+    ///
+    /// 1. Identify the indices of all instance columns in the permutation and collect them in `instance_columns`.
+    /// 2. Filter out and sort the indices to be removed that match the instance columns into `columns_to_remove_set`.
+    /// 3. Iterate through each column in the permutation:
+    ///    - If the column is in `columns_to_remove_set`:
+    ///      - Reset its permutation by setting each of its mapping cells to the identity (i.e., `(column_index, row_index)`).
+    ///    - Otherwise, for each row in the column:
+    ///      - Traverse the cycle starting from the current cell. If a cycle includes a row in the columns to remove, bypass that row.
+    ///      - Update the mapping to link directly to the next cell in the cycle that is not in the columns to remove.
+    /// 4. Return the updated `PermutationData` object.
+    pub(crate) fn rm_copy_constraints(
+        mut self,
+        instance_columns_to_remove: impl Iterator<Item = usize>,
+    ) -> Self {
+        let instance_columns = self
+            .columns
+            .iter()
+            .filter(|column| column.column_type().eq(&Any::Instance))
+            .map(|column| column.index())
+            .collect::<Box<[_]>>();
+
+        let mut columns_to_remove_set = instance_columns_to_remove
+            .filter(|index| instance_columns.binary_search(index).is_ok())
+            .collect::<Box<[_]>>();
+
+        columns_to_remove_set.sort();
+
+        for (column_index, column) in self.columns.iter().enumerate() {
+            if columns_to_remove_set.binary_search(&column.index()).is_ok() {
+                debug!("completely clearing all permutations for column {column:?}");
+                for (row_index, mapping_cell) in self.mapping[column_index].iter_mut().enumerate() {
+                    *mapping_cell = (column_index, row_index);
+                }
+            } else {
+                let row_count = self.mapping[column_index].len();
+
+                for row_index in 0..row_count {
+                    let (mut next_i, mut next_j) = self.mapping[column_index][row_index];
+
+                    let start = (column_index, row_index);
+                    while (next_i, next_j) != start
+                        && columns_to_remove_set
+                            .binary_search(&self.columns[next_i].index())
+                            .is_ok()
+                    {
+                        (next_i, next_j) = if (next_i, next_j) == self.mapping[next_i][next_j] {
```

Comment:
Let's say columns to be removed is "bad" column, columns not to be removed is good column.

(1) while in practice it usually will not happen, but it is still possible that both (next_i, next_j) and self.mapping[next_i][next_j] are in bad columns. This need to be considered.

(2) e.g. suppose we have cycle like this:
(i0,j0)->(i1,j1)->(i2,j2)->(i0,j0)
where i0 is good column and i1 is bad column. what we want is to change it to: (i0,j0)->(i2,j2)->(i0,j0), but since you already set self.maping[i1][j1]=(i1,j1), now it will change the original cycle to (i0,j0)->(i0,j0)  (correct me if I am wrong)

It seems we should update the cycles that starting from "good" columns, then update cycles that starting from "bad" columns.

Reply:
Let me fix it in next PRs, for easy merge all of it